File: 2023/2/solution.js
const fs = require("fs");
const path = require("path");

const input = fs
  .readFileSync(path.join(__dirname, "input.txt"), "utf-8", (err, data) => {
    if (err) {
      console.log(err);
    } else {
      return data;
    }
  })
  .split("\n");

const red = 12;
const green = 13;
const blue = 14;

function Game(record) {
  const parts = record.split(": ");
  this.id = Number(parts[0].split(" ")[1]);
  this.subsets = [];
  for (const subsetString of parts[1].split("; ")) {
    const subset = new Map();
    for (const cubes of subsetString.split(", ")) {
      const parts2 = cubes.split(" ");
      subset.set(parts2[1], Number(parts2[0]));
    }
    this.subsets.push(subset);
  }
}

const games = [];

for (const line of input) {
  games.push(new Game(line));
}

let part1 = 0;
let part2 = 0;

for (const game of games) {
  let isPossible = true;
  let maxRed = 0;
  let maxGreen = 0;
  let maxBlue = 0;
  for (const subset of game.subsets) {
    if (
      isPossible &&
      ((subset.get("red") ?? 0) > red ||
        (subset.get("green") ?? 0) > green ||
        (subset.get("blue") ?? 0) > blue)
    ) {
      isPossible = false;
    }
    maxRed = Math.max(maxRed, subset.get("red") ?? 0);
    maxGreen = Math.max(maxGreen, subset.get("green") ?? 0);
    maxBlue = Math.max(maxBlue, subset.get("blue") ?? 0);
  }
  if (isPossible) {
    part1 += game.id;
  }
  part2 += maxRed * maxGreen * maxBlue;
}

console.log(part1);
console.log(part2);

File: 2023/2/solution.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"strconv"
	"strings"
)

func main() {
	_, filename, _, _ := runtime.Caller(0)
	dirname := filepath.Dir(filename)
	inputFilePath := filepath.Join(dirname, "input.txt")
	data, err := os.ReadFile(inputFilePath)
	if err != nil {
		panic(err)
	}
	games := parseInput(strings.Split(string(data), "\n"))
	fmt.Println(part1(games))
	fmt.Println(part2(games))
}

var (
	numRegex  = regexp.MustCompile(`\d+`)
	cubeRegex = regexp.MustCompile(`(\d+) (\w+)`)
	colors    = []string{"red", "green", "blue"}
)

type Game struct {
	id            int
	revealedCubes []map[string]int
}

func (g *Game) IsPossible(testBag map[string]int) bool {
	for _, cubeSet := range g.revealedCubes {
		for col, n := range testBag {
			if cubeSet[col] > n {
				return false
			}
		}
	}
	return true
}

func (g *Game) MinimumSet() map[string]int {
	minSet := make(map[string]int)
	for _, cubeSet := range g.revealedCubes {
		for col, n := range cubeSet {
			if n > minSet[col] {
				minSet[col] = n
			}
		}
	}
	return minSet
}

func PowerOfSet(set map[string]int) int {
	res := 1
	for _, col := range colors {
		res *= set[col]
	}
	return res
}

func NewGame(line string) *Game {
	parts := strings.Split(line, ":")
	cubeSets := strings.Split(parts[1], ";")
	revealedCubes := make([]map[string]int, len(cubeSets))
	match := numRegex.FindString(parts[0])
	id, err := strconv.Atoi(match)
	if err != nil {
		panic(err)
	}
	for i, set := range cubeSets {
		matches := cubeRegex.FindAllStringSubmatch(set, -1)
		cubes := make(map[string]int, len(matches))
		for _, match := range matches {
			n, err := strconv.Atoi(match[1])
			if err != nil {
				panic(err)
			}
			cubes[match[2]] = n
		}
		revealedCubes[i] = cubes
	}
	return &Game{id, revealedCubes}
}

func parseInput(data []string) []*Game {
	games := make([]*Game, len(data))
	for i, line := range data {
		games[i] = NewGame(line)
	}
	return games
}

func part1(games []*Game) int {
	res := 0
	testBag := map[string]int{"red": 12, "green": 13, "blue": 14}
	for _, game := range games {
		if game.IsPossible(testBag) {
			res += game.id
		}
	}
	return res
}

func part2(games []*Game) int {
	res := 0
	for _, game := range games {
		res += PowerOfSet(game.MinimumSet())
	}
	return res
}
